src/schemapack/utils.py

```diff
@@ -108,3 +114,33 @@ def transient_directory_change(path: Path):
         yield
     finally:
         os.chdir(original_cwd)
+
+
+_K = TypeVar("_K")
+_V_co = TypeVar("_V_co", covariant=True)
+
+
+class FrozenDict(immutabledict[_K, _V_co]):
+    """A pydantic-comatible wrapper around immutabledict."""
+
+    @classmethod
+    def __get_pydantic_core_schema__(
+        cls, source: Any, handler: GetCoreSchemaHandler
+    ) -> core_schema.CoreSchema:
+        """Get the pydantic core schema for this type."""
+        # Validate the type against a dict:
+        # (this will have the side effect of converting the instance to dict even if it
+        # is already a dict a FrozenDict)
+        args = typing.get_args(source)
+        if args:
+            if len(args) != 2:
+                raise TypeError(
+                    "Expected exactly two (or no) type arguments for FrozenDict, got"
+                    + f" {len(args)}"
+                )
+            dict_schema = handler.generate_schema(dict[args[0], args[1]])  # type: ignore
```

Comment:
```suggestion
            if len(args) == 2:
                dict_schema = handler.generate_schema(dict[args[0], args[1]])  # type: ignore
            else:
                raise TypeError(
                    "Expected exactly two (or no) type arguments for FrozenDict, got"
                    + f" {len(args)}"
                )
            
```
Is a little easier for me to read. Feel free to ignore.

Reply:
Hi Byron, I agree that positive comparisons are easier to understand than negative ones. At the same time, it is a common python pattern that you try to avoid the use of else and avoid inserting branches into your main (non-error) flow. In this case where there is only one error, it is not so severe but consider this example:

```
# do something

if ...: # make an assertion
    # raise an exception if assertion does not hold

# continue main code flow

if ...: # make another assertion
    # raise an exception if assertion does not hold

# continue main code flow
```

Compare this to:
```
# do something

if ...: # make an assertion
    # continue main code flow
    
    if ...: # make another assertion
        # continue main code flow
    else:
        # raise an exception

else:
    # raise an exception
```

@Cito do you know whether this python pattern has a name?

Generally, this code would also be a good candidate for pattern matching which is, however, only available starting with 3.10.

Reply:
This might be a slight improvement (at least with regard to nesting):

```
        args = typing.get_args(source)
        if not args:
            dict_schema = handler.generate_schema(dict)
        elif len(args) == 2:
            dict_schema = handler.generate_schema(dict[args[0], args[1]])  # type: ignore
        else:
            raise TypeError(
                "Expected exactly two (or no) type arguments for FrozenDict, got"
                + f" {len(args)}"
            )
        
        # Uses cls as validator function to convert the dict to a FrozenDict:
        return core_schema.no_info_after_validator_function(cls, dict_schema)
```
Have implemented this in the latest commit.

Reply:
Thanks

Reply:
Not sure if it has a name, but I think the first style is more in line with the Zen of Python "Flat is better than nested." It's a bit related to the "return early" pattern.

Structural pattern matching is one good reason to upgrade to Python > 3.10.